### PR TITLE
Optimize memory usage of Softmax

### PR DESCRIPTION
**Pull request by kakao-steve-ai**
## 🙏 Describe the pull request

* 큰 텐서의 경우 Backward 안되던 문제 해결했습니다.
* 메모리 많이 사용하는 diagonal 행렬 대신 메모리 효율적인 방법을 사용했습니다. 
* dim=0, 1 지원하도록 수정했습니다.

## ✅ Checklist

- [x] Code follows the project's coding conventions and style.
- [x] Tests have been added or updated to cover the changes.
- [x] Documentation has been updated, if necessary.

### CODE REVIEW

**Comment by daemyung** [comment]
@ansteve 고생하셨습니다. 커밋 메세지를 다음 중에서 마음에 드는걸로 변경해주실 수 있을까요?

1. backward의 첫 글자를 소문자로 변경했습니다.
```
Optimize Softmax backward
```

2. backward를 강조, 하지만 1번이 더 깔끔합니다.
```
Optimize the backward of Softmax
```

3. 메모리 사용량 최적화를 강조합니다.
```
Optimize memory usage of Softmax
    
Only tensors of small sizes were able to be backpropagated.
```

**Comment by kakao-steve-ai** [comment]
> @ansteve 고생하셨습니다. 커밋 메세지를 다음 중에서 마음에 드는걸로 변경해주실 수 있을까요?
> 

3번으로 수정했습니다. 감사합니다.

File: tests/test_softmax.py [hunk 1]
```diff
@@ -19,28 +19,28 @@
...
     input = torch.randn(y_size, x_size, **factory_kwargs)
 
     assert util.equal(torch.nn.functional.softmax(input, dim), trident.function.softmax(input, dim))
 
 
-@pytest.mark.parametrize("num_vec, vec_sz", [(4, 64), (5, 70)])
-def test_backward(num_vec, vec_sz, device, dtype):
+@pytest.mark.parametrize("y_size, x_size, dim", [(300, 500, 0), (5, 7000, 1)])
+def test_backward(y_size, x_size, dim, device, dtype):
     ctor_args = {"device": device, "dtype": dtype}
```
Comment (mejai1206) [inline tests/test_softmax.py]: 변수명을 풀어쓰기로 한 뒤로는
`ctor_args` 대신 `factory_kwargs` 로 쓰이는 코드가 자주 보이는데, `factory_kwargs`로의 변경 괜찮을까요?

Comment (kakao-steve-ai) [reply tests/test_softmax.py]: factory_kwargs 로 변경하겠습니다.

File: tests/test_softmax.py [hunk 2]
```diff
@@ -19,28 +19,28 @@
...
 
-@pytest.mark.parametrize("num_vec, vec_sz", [(4, 64), (5, 70)])
-def test_backward(num_vec, vec_sz, device, dtype):
-    ctor_args = {"device": device, "dtype": dtype}
-    inp = torch.randn(num_vec, vec_sz, **ctor_args)
-    tgt = torch.randn(num_vec, vec_sz, **ctor_args)
+@pytest.mark.parametrize("y_size, x_size, dim", [(300, 500, 0), (5, 7000, 1)])
+def test_backward(y_size, x_size, dim, device, dtype):
+    factory_kwargs = {"device": device, "dtype": dtype}
+    inp = torch.randn(y_size, x_size, **factory_kwargs)
```
Comment (mejai1206) [inline tests/test_softmax.py]: `inp` 와 `tgt`를  풀어서 쓰는 것 괜찮을까요?

Comment (kakao-steve-ai) [reply tests/test_softmax.py]: 수정하겠습니다

File: trident/operation/softmax.py [hunk 3]
```diff
@@ -59,13 +56,39 @@ def grid(meta):
 
     @staticmethod
     def __backward(grad_output: torch.Tensor, output: torch.Tensor, dim: int):
-        y_size, x_size = output.shape
+        if dim == 0:
```
Comment (mejai1206) [inline trident/operation/softmax.py:59]: 티타임에서 얘기주신  `dim`에 따른 `block_pointer` 생성 분기를 없애주신 방법은 여기인걸까요?

**Review by mejai1206 — APPROVED** [review]
LGTM

File: trident/operation/softmax.py (hunk 3 above)
Comment (kakao-steve-ai) [reply trident/operation/softmax.py:59]: 네 이런식으로 하면 커널에서 분기를 없앨 수 있습니다. 분기가 없어지는 대신 약간 머릿속이 복잡해지는 것 같고, 어떤게 더 좋을지 모르겠네요.

Comment (mejai1206) [reply trident/operation/softmax.py:59]: 아 넵 전 이 방법 좋은 것 같습니다~

File: trident/kernel/softmax.py [hunk 4]
```diff
@@ -18,7 +18,7 @@
 from trident import kernel, language
 
 
-def configs_for_forward():
+def configs_for_softmax():
```
Comment (daemyung) [inline trident/kernel/softmax.py]: `configs`로도 충분하다고 생각합니다.

File: trident/operation/softmax.py [hunk 5]
```diff
@@ -59,13 +56,39 @@ def grid(meta):
...
+            x_size, y_size = output.shape
+            stride_0 = 1
```
Comment (daemyung) [inline trident/operation/softmax.py]: 이런 경우에 `size_along_dim`이라는 변수 이름을 사용하고 있습니다. 다른 `stride_1`은 항상 1이니까 선언하지 않고요. 그리고 이 계산을 커널 안으로 옮기는게 더 좋을 것 같습니다.

File: trident/operation/softmax.py [hunk 6]
```diff
@@ -59,13 +56,39 @@ def grid(meta):
...
+            stride_1 = 1
         grad_input = torch.empty_like(output)
-        block_size = max(triton.next_power_of_2(x_size), 16)
 
         def grid(meta):
-            return (x_size if dim == 0 else y_size,)
+            return (y_size,)
 
-        kernel.Softmax.backward[grid](grad_output, output, x_size, grad_input, block_size)
+        delta = torch.empty(y_size, dtype=torch.float32, device="cuda")
```
Comment (daemyung) [inline trident/operation/softmax.py]: `delta` 생성시 `factory_kwrags = {"device" = grad_output.device, "dtype" = grad_output.dtype"}`을 정의하고 생성해 주세요.

File: trident/kernel/softmax.py [hunk 7]
```diff
@@ -126,22 +126,118 @@ def forward(
             input_block_ptr = tl.advance(input_block_ptr, (0, block_size))
             output_block_ptr = tl.advance(output_block_ptr, (0, block_size))
 
-    @staticmethod
+    @triton.autotune(
+        configs=configs_for_softmax(),
+        key=["y_size", "x_size"],
+    )
+    @triton.jit
+    def backward_preprocess(
```
Comment (daemyung) [inline trident/kernel/softmax.py]: `backward_delta`로 이름 변경 제안드립니다. `input`의 역전파를 제외하고는 `backward_{변수이름}`으로 함수 이름을 만들 고 있습니다.

**Review by daemyung — CHANGES_REQUESTED** [review]

File: trident/kernel/softmax.py (hunk 4 above)
Comment (kakao-steve-ai) [reply trident/kernel/softmax.py]: `trident/kernel/__init__.py` 에서 `import *` 를 해주고 있어서, 나중에 다른 파일에 있는 `configs`와 이름 충돌이 생길까봐 이렇게 했습니다. 어떻게 하는게 좋을까요?

Comment (kakao-steve-ai) [reply trident/kernel/softmax.py]: 이름을 `configs` 로 바꾸고 `Softmax` 클래스 안쪽으로 이동시키면 되겠네요.

Comment (daemyung) [reply trident/kernel/softmax.py]: 이름 충돌이 생겨도 어짜피 `Softmax`만 사용하는 거라 상관 없다고 생각합니다. 클래스 안으로 정의하는 것은 안되더라고요?

File: trident/operation/softmax.py (hunk 6 above)
Comment (kakao-steve-ai) [reply trident/operation/softmax.py]: 중간 계산 과정에서는 `float32`를 사용하고 있어서, 굳이 `grad_output.dtype` 정의해서 변환하는 것이 불필요해보였습니다. `factory_kwargs`는 변경하겠습니다.

File: trident/kernel/softmax.py (hunk 7 above)
Comment (kakao-steve-ai) [reply trident/kernel/softmax.py]: 이름 수정하겠습니다

File: trident/operation/softmax.py (hunk 6 above)
Comment (daemyung) [reply trident/operation/softmax.py]: 아하! 그런 이유라면 dtype이 필요 없어보입니다.

File: trident/operation/softmax.py (hunk 5 above)
Comment (kakao-steve-ai) [reply trident/operation/softmax.py]: `x_size`, `y_size`를 커널에 넘겨주기 보다 `size_along_dim`, `size_against_dim`라는 이름으로 넘겨주는게 좋아보이네요. `size_against_dim`라는 이름 괜찮아보이시나요?

File: trident/kernel/softmax.py (hunk 4 above)
Comment (kakao-steve-ai) [reply trident/kernel/softmax.py]: 클래스 안으로 넣는거 되는거 같습니다.

File: trident/operation/softmax.py (hunk 5 above)
Comment (kakao-steve-ai) [reply trident/operation/softmax.py]: 커널 안에서 `strides = (1, size_against_dim) if dim == 0 else (size_along_dim, 1)` 이런식으로 변수 선언하니까 컴파일 에러가 발생해서 일단 stride_0, stride_1 라는 변수 유지했습니다.

File: trident/kernel/softmax.py [hunk 8]
```diff
@@ -127,21 +126,129 @@ def forward(
...
     @staticmethod
+    @triton.autotune(
+        configs=configs(),
+        key=["size_against_dim", "size_along_dim"],
+    )
+    @triton.jit
+    def backward_delta(
+        grad_output_ptr: tl.tensor,
+        output_ptr: tl.tensor,
+        delta_ptr: tl.tensor,
```
Comment (daemyung) [inline trident/kernel/softmax.py]: 리턴값을 아웃 파라미터로 정의하는 경우에 리턴값부터 아규먼트로 선언됩니다. `delta_ptr`을 제일 앞으로 옮겨주세요.

File: trident/operation/softmax.py (hunk 5 above)
Comment (daemyung) [reply trident/operation/softmax.py]: @ansteve 이름 좋습니다.

File: trident/kernel/softmax.py [hunk 9]
```diff
@@ -127,21 +126,129 @@ def forward(
...
+        size_against_dim: int,
+        size_along_dim: int,
+        dim: int,
+        block_size: tl.constexpr,
+    ):
+        offset = tl.program_id(0)
+        if dim == 0:
+            stride_0 = 1
+            stride_1 = size_against_dim
```
Comment (daemyung) [inline trident/kernel/softmax.py]: `size_against_dim`, `size_along_dim`을 내부적으로 계산하면 안되는 건가요?

File: trident/kernel/softmax.py [hunk 10]
```diff
@@ -127,21 +126,129 @@ def forward(
...
+        grad_output_block_ptr = tl.make_block_ptr(
+            grad_output_ptr,
+            shape=(size_against_dim, size_along_dim),
+            strides=(stride_0, stride_1),
+            offsets=(offset, 0),
+            block_shape=(1, block_size),
+            order=(1, 0),
+        )
+
+        delta = tl.zeros((1,), tl.float32)
```
Comment (daemyung) [inline trident/kernel/softmax.py]: 최적화가 필요해 보입니다.
1. delta를 block_size의 크기로 선언해주세요.

File: trident/kernel/softmax.py [hunk 11]
```diff
@@ -127,21 +126,129 @@ def forward(
...
+        for _ in range(0, size_along_dim, block_size):
+            output = tl.load(output_block_ptr, boundary_check=(1,), padding_option="zero").to(tl.float32)
+            grad_output = tl.load(grad_output_block_ptr, boundary_check=(1,), padding_option="zero").to(tl.float32)
+            delta += tl.sum(output * grad_output, axis=1)
```
Comment (daemyung) [inline trident/kernel/softmax.py]: 최적화 연속 커멘트입니다.
2. `delta += output * grad_output`로 더해주세요.

File: trident/kernel/softmax.py [hunk 12]
```diff
@@ -127,21 +126,129 @@ def forward(
...
+            output_block_ptr = tl.advance(output_block_ptr, (0, block_size))
```
Comment (daemyung) [inline trident/kernel/softmax.py:177]: 최적화 연속 커멘트입니다.
3. `delta = tl.sum(delta, 1)`로 최종 delta를 구해주세요.

이렇게 하면 Reduce 횟수를 줄일 수 있습니다.

**Review by daemyung — CHANGES_REQUESTED** [review]

File: trident/operation/softmax.py [hunk 13]
```diff
@@ -60,12 +57,37 @@ def grid(meta):
...
+        if dim == 0:
+            size_along_dim = y_size
+            size_against_dim = x_size
+        else:
+            size_along_dim = x_size
+            size_against_dim = y_size
 
         def grid(meta):
-            return (x_size if dim == 0 else y_size,)
+            return (size_against_dim,)
```
Comment (kakao-steve-ai) [inline trident/operation/softmax.py]: > `size_against_dim`, `size_along_dim`을 내부적으로 계산하면 안되는 건가요?

여기와 아래 line 71에서 사용하고 싶어서 커널 외부에서 설정했습니다.

File: trident/kernel/softmax.py (hunk 12 above)
Comment (kakao-steve-ai) [reply trident/kernel/softmax.py:177]: 아하 수정하겠습니다.

File: trident/operation/softmax.py (hunk 13 above)
Comment (daemyung) [reply trident/operation/softmax.py]: 아 네, 제 말은 `size_against_dim`, `size_along_dim`을 커널로 넘기지 말고 커널에서 다시 구해주는게 어떠냐는 의미였어요.

Comment (kakao-steve-ai) [reply trident/operation/softmax.py]: 아 그런뜻이었군요. 외부에서 이미 구했으니 넘겨주는것도 나쁘지 않은거 같습니다.

Comment (daemyung) [reply trident/operation/softmax.py]: 내부에서 계산해 주시겠어요? 커널에 파라미터를 많이 넘기는게 성능상으로 좋지 않거든요.

Comment (kakao-steve-ai) [reply trident/operation/softmax.py]: 넵

Comment (kakao-steve-ai) [reply trident/operation/softmax.py]: 커널 안에서는 `output.shape`를 가져올 수가 없네요. 결국 `x_size`, `y_size `를 넘겨주거나 `size_against_dim`, `size_along_dim` 를 넘겨주어야하는데, 이렇게 되면 커널에 넘겨줘야하는 파라미터 개수는 똑같아서 그대로 둬도 괜찮을거 같습니다.

Comment (daemyung) [reply trident/operation/softmax.py]: 네 맞습니다. `x_size`, `y_size`를 넘겨주고 커널에서 `size_along_dim`, `size_against_dim`을 계산해주세요. 지금 코딩 스타일이 이런데 티타임에 이야기해서 어느것을 넘기는게 좋을지 이야기해봐요.

Comment (kakao-steve-ai) [reply trident/operation/softmax.py]: 넵

File: trident/kernel/softmax.py [hunk 14]
```diff
@@ -127,21 +126,139 @@ def forward(
...
+        x_size: int,
+        dim: int,
+        block_size: tl.constexpr,
+    ):
+        offset = tl.program_id(0)
+
+        if dim == 0:
+            size_along_dim = y_size
+            size_against_dim = x_size
+            stride_0 = 1
```
Comment (daemyung) [inline trident/kernel/softmax.py]: 이렇게 하더라도 stride_0, stride_1이 필요한가요?

Comment (kakao-steve-ai) [reply trident/kernel/softmax.py]: 이것도 코딩스타일 문제인거 같습니다. 아래쪽 `make_block_ptr`에서 반복적으로 `strides = (1, size_against_dim) if dim == 0 else (size_along_dim, 1)` 하는것보다 이 방식이 좋다고 생각했습니다. 다른 옵션으로, 새로운 변수 선언하고 그것을 `make_block_ptr`에서 사용하려고 했을때는 컴파일 에러가 발생했고요.

Comment (daemyung) [reply trident/kernel/softmax.py]: 이거 내일 티타임에 이야기해서 정리하시죠.

Comment (kakao-steve-ai) [reply trident/kernel/softmax.py]: `x_size`, `y_size`, `x_stride`, `y_stride` 로 하는게 제일 깔끔한거 같습니다.

**Review by daemyung — APPROVED** [review]
LGTM, 고생하셨습니다.

File: trident/kernel/softmax.py (hunk 14 above)
Comment (daemyung) [reply trident/kernel/softmax.py]: 저도 그렇게 생각됩니다. 이 방법을 정착 시키죠.
